File: app/src/main/java/com/shentu/wallpaper/model/api/service/MicroService.java
package com.shentu.wallpaper.model.api.service;


import com.shentu.wallpaper.model.entity.MicroUser;
import com.shentu.wallpaper.model.response.BannerPageResponse;
import com.shentu.wallpaper.model.response.BaseResponse;
import com.shentu.wallpaper.model.response.CategoryPageResponse;
import com.shentu.wallpaper.model.response.SplashAdResponse;
import com.shentu.wallpaper.model.response.SubjectDetailResponse;
import com.shentu.wallpaper.model.response.SubjectPageResponse;
import com.shentu.wallpaper.model.response.WallpaperPageResponse;

import io.reactivex.Observable;
import retrofit2.http.Body;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.PATCH;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface MicroService {

    int PAGE_LIMIT = 20;
    int PAGE_START = 0;

    @GET("splash")
    Observable<SplashAdResponse>
    getSplash();

    @GET("subjects/{pk}")
    Observable<SubjectDetailResponse>
    getSubjectDetail(@Path("pk") Integer pk);

    @GET("wallpapers")
    Observable<WallpaperPageResponse>
    getSubjectWallpapers(@Query("subject_id") int subjectId,
                         @Query("limit") int limit,
                         @Query("offset") int offset);

    @GET("wallpapers")
    Observable<WallpaperPageResponse>
    getBannerWallpapers(@Query("banner_id") int subjectId,
                        @Query("limit") int limit,
                        @Query("offset") int offset);

    @GET("wallpapers")
    Observable<WallpaperPageResponse>
    getCategoryWallpapers(@Query("category_id") int categoryId,
                          @Query("limit") int limit,
                          @Query("offset") int offset);

    /**
     * 获取推荐的WallPaper
     */
    @GET("recommend")
    Observable<WallpaperPageResponse>
    getRecommendWallpapers(@Query("limit") int limit,
                           @Query("offset") int offset);

    @GET("categories")
    Observable<CategoryPageResponse>
    getCategories(@Query("offset") int offset,
                  @Query("limit") int limit);

    @FormUrlEncoded
    @POST("account/register/")
    Observable<BaseResponse<Boolean>>
    getRegisterAccount(@Field("phone") String phone,
                       @Field("password") String password);

    @FormUrlEncoded
    @POST("account/login/")
    Observable<BaseResponse<MicroUser>>
    loginAccount(@Field("phone") String phone);

    @POST("account/logout/")
    Observable<BaseResponse<String>>
    logout();

    @GET("subjects/")
    Observable<SubjectPageResponse>
    searchKey(@Query("key") String key,
              @Query("limit") int limit,
              @Query("offset") int offset);

    @GET("banners/")
    Observable<BannerPageResponse>
    getBanners(@Query("limit") int limit,
               @Query("offset") int offset);

    @FormUrlEncoded
    @POST("category/update/")
    Observable<BaseResponse<Boolean>>
    updateCategoryCover(@Field("cid") Integer cid, @Field("logo") String logo);

    @PATCH("user/update/{pk}")
    Observable<MicroUser>
    updateUser(@Body MicroUser user, @Path("pk") int pk);

    @POST("buy/paper/{pk}")
    Observable<BaseResponse<String>>
    buyPaper(@Path("pk") int pk, @Query("pea") int pea);

//    /**
//     * collect
//     * */
//    @GET("collect/my")
//    Observable<WallpaperPageResponse>
//    getMyCollects(@Query("offset") int offset);
}

File: app/src/main/java/com/shentu/wallpaper/model/api/service/MicroService.kt
package com.shentu.wallpaper.model.api.service

import com.shentu.wallpaper.model.entity.AppUpdate
import com.shentu.wallpaper.model.entity.MicroUser
import com.shentu.wallpaper.model.entity.Wallpaper
import com.shentu.wallpaper.model.response.*
import io.reactivex.Observable
import retrofit2.http.*

interface MicroService {
    @get:GET("splash")
    val splash: Observable<SplashAdResponse>

    @get:GET("update")
    val updateInfo: Observable<BaseResponse<AppUpdate>>

    @GET("subjects/{pk}")
    fun getSubjectDetail(@Path("pk") pk: Int): Observable<SubjectDetailResponse>

    @GET("wallpapers")
    fun getSubjectWallpapers(@Query("subject_id") subjectId: Int,
                             @Query("limit") limit: Int,
                             @Query("offset") offset: Int): Observable<WallpaperPageResponse>

    @GET("wallpapers")
    fun getBannerWallpapers(@Query("banner_id") subjectId: Int,
                            @Query("limit") limit: Int,
                            @Query("offset") offset: Int): Observable<WallpaperPageResponse>

    @GET("wallpapers")
    fun getCategoryWallpapers(@Query("category_id") categoryId: Int,
                              @Query("limit") limit: Int,
                              @Query("offset") offset: Int): Observable<WallpaperPageResponse>

    /**
     * 获取推荐的WallPaper
     */
    @GET("recommend")
    fun getRecommendWallpapers(@Query("limit") limit: Int,
                               @Query("offset") offset: Int): Observable<WallpaperPageResponse>

    /**
     * 获取最新papers
     */
    @GET("wallpaper/newest")
    fun getNewestPapers(@Query("limit") limit: Int,
                        @Query("offset") offset: Int): Observable<WallpaperPageResponse>

    @GET("categories")
    fun getCategories(@Query("offset") offset: Int,
                      @Query("limit") limit: Int): Observable<CategoryPageResponse>

    @FormUrlEncoded
    @POST("account/register/")
    fun getRegisterAccount(@Field("phone") phone: String,
                           @Field("password") password: String): Observable<BaseResponse<Boolean>>

    @FormUrlEncoded
    @POST("account/login/")
    fun loginAccount(@Field("phone") phone: String): Observable<BaseResponse<MicroUser>>

    @POST("account/logout/")
    fun logout(): Observable<BaseResponse<String>>

    @GET("subjects/")
    fun searchKey(@Query("key") key: String,
                  @Query("limit") limit: Int,
                  @Query("offset") offset: Int): Observable<SubjectPageResponse>

    @GET("banners/")
    fun getBanners(@Query("limit") limit: Int,
                   @Query("offset") offset: Int): Observable<BannerPageResponse>

    @FormUrlEncoded
    @POST("category/update/")
    fun updateCategoryCover(@Field("cid") cid: Int,
                            @Field("logo") logo: String): Observable<BaseResponse<Boolean>>

    @FormUrlEncoded
    @POST("paper/set/banner")
    fun addPaper2Banner(@Field("bid") bid: Int,
                         @Field("pid") pid: Int): Observable<BaseResponse<Boolean>>

    @PATCH("user/update/{pk}")
    fun updateUser(@Body user: MicroUser, @Path("pk") pk: Int): Observable<MicroUser>

    @POST("buy/paper/{pk}")
    fun buyPaper(@Path("pk") pk: Int,
                 @Query("type") type: Int): Observable<BaseResponse<Int>>

    @GET("wallpaper/detail/{pk}")
    fun getPaperDetail(@Path("pk") pk: Int): Observable<BaseResponse<Wallpaper>> //    /**

    //     * collect
//     * */
//    @GET("collect/my")
//    Observable<WallpaperPageResponse>
//    getMyCollects(@Query("offset") int offset);
    companion object {
        const val PAGE_LIMIT = 20
        const val PAGE_START = 0
    }
}